File: cpp/leet_code/1992.cpp
/**
 * @file 1992.cpp
 * @brief Contains the implementation of the Solution class.
 *
 * This file contains the implementation of a Solution class which provides a method to find farmland in a given land matrix.
 */

#include <ios> // For std::ios_base::sync_with_stdio
#include <iostream> // For std::cin.tie
#include <vector> // For std::vector

/**
 * @class Solution
 * @brief A class providing a method to find farmland in a given land matrix.
 *
 * This class provides a method to find farmland in a given land matrix. The farmland is represented by a 1 in the matrix and
 * each farmland is a rectangle. The method returns the top-left and bottom-right coordinates of each farmland.
 */
class Solution {
public:
    /**
     * @brief Finds the farmland in a given land matrix.
     *
     * This method finds the farmland in a given land matrix. The farmland is represented by a 1 in the matrix and each farmland is a rectangle.
     * The method returns the top-left and bottom-right coordinates of each farmland.
     *
     * @param land The land matrix.
     * @return A vector of vectors where each inner vector contains the top-left and bottom-right coordinates of a farmland.
     */
    std::vector<std::vector<int> > findFarmland(std::vector<std::vector<int> > &land) {
        std::ios_base::sync_with_stdio(false); // Speeds up I/O operations
        std::cin.tie(nullptr); // Unties cin from cout

        const int n = land.size(); // The number of rows in the land matrix
        const int m = land[0].size(); // The number of columns in the land matrix
        std::vector<std::vector<int> > result; // The result vector

        // Iterate over the land matrix
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (!land[i][j]) continue; // If the current cell is not a farmland, continue to the next cell

                int c2 = j;
                // Find the right boundary of the farmland
                while (c2 < m && land[i][c2]) {
                    c2++;
                }
                int r2 = i;
                // Find the bottom boundary of the farmland
                while (r2 < n && land[r2][j]) {
                    r2++;
                }
                c2 = !c2 ? c2 : c2 - 1;
                r2 = !r2 ? r2 : r2 - 1;
                result.push_back({i, j, r2, c2}); // Add the farmland to the result

                // Mark the cells of the farmland as visited
                for (int k = i; k <= r2; k++) {
                    for (int l = j; l <= c2; l++) {
                        land[k][l] = 0;
                    }
                }
            }
        }
        return result; // Return the result
    }
};

File: php/leet_code/1992.php
<?php

/**
 * This file contains the Solution class.
 *
 * PHP version 8.3
 *
 * @category LeetCode
 * @package  LeetCode\Solution
 * @author   Jerodg
 * @license  http://opensource.org/licenses/gpl-license.php GNU Public License
 * @link     http://github.com/jerodg/leet_code
 */

declare(strict_types=1);

/**
 * The Solution class contains methods for solving specific problems related to farmland.
 *
 * @category LeetCode
 * @package  LeetCode\Solution
 * @author   Jerodg
 * @license  http://opensource.org/licenses/gpl-license.php GNU Public License
 * @link     http://github.com/jerodg/leet_code
 */
class Solution
{
    /**
     * Finds the farmland in a given land matrix.
     *
     * This method takes a 2D array representing a land matrix, where 1 represents farmland and 0 represents non-farmland.
     * It returns an array of farmlands, where each farmland is represented as an array [i, j, x, y] where (i, j) is the top-left corner of the farmland and (x, y) is the bottom-right corner of the farmland.
     *
     * @param array $land The land matrix.
     *
     * @return array The array of farmlands.
     *
     * @throws InvalidArgumentException If $land is not a 2D array.
     *
     * @example
     * $solution = new Solution();
     * $farmlands = $solution->findFarmland([[1,0,0],[0,1,1],[0,1,1]]);
     * print_r($farmlands); // Outputs: [[0,0,0,0],[1,1,2,2]]
     */
    public function findFarmland(array $land): array
    {
        // Initialize the output array
        $output = [];

        // Calculate the number of columns in the land matrix
        $numColumns = count($land[0]);

        // Iterate over each row in the land matrix
        foreach ($land as $i => $iValue) {

            // Iterate over each column in the current row
            for ($j = 0; $j < $numColumns; $j++) {

                // Get the current cell value
                $n = $iValue[$j];

                // Check if the current cell is the top-left corner of a farmland
                if ($n && ($i == 0 || $land[$i - 1][$j] == 0) && ($j == 0 || $iValue[$j - 1] == 0)) {

                    // Initialize the bottom-right corner of the farmland
                    $x = $i;
                    $y = $j;

                    // Find the bottom boundary of the farmland
                    while ($x < count($land) - 1 && $land[$x + 1][$y] == 1) {
                        $x++;
                    }

                    // Find the right boundary of the farmland
                    while ($y < $numColumns - 1 && $land[$x][$y + 1] == 1) {
                        $y++;
                    }

                    // Add the farmland to the output array
                    $output[] = [$i, $j, $x, $y];
                }
            }
        }

        // Return the output array
        return $output;
    }
}
